src/System.Private.CoreLib/shared/System/ThrowHelper.ttinclude

```diff
@@ -0,0 +1,110 @@
+﻿<#+
+
+/*
+ * Don't make changes to this file.
+ * Make changes to ExceptionArgument.cs / ExceptionResource.cs instead.
+ */
+
+//
+// Writes "internal enum ENUMNAME { ... }"
+//
+
+void WriteEnumType(string enumName, Type enumType)
+{
+#>
+    internal enum <#= enumName #>
+    {
+<#+
+foreach (string value in Enum.GetNames(enumType))
+{
+#>
+        <#= value #>,
+<#+
+}
+#>
+    }
+<#+
+}
+
+//
+// Writes ThrowHelper.GetArgumentName
+//
+
+void WriteThrowHelperGetArgumentName(Type exceptionArgumentType)
+{
+#>
+    internal static partial class ThrowHelper
+    {
+        private static string GetArgumentName(ExceptionArgument argument)
+        {
+            switch (argument)
+            {
+<#+
+    foreach (string enumName in Enum.GetNames(exceptionArgumentType))
+    {
+#>
+                case ExceptionArgument.<#= enumName #>:
+                    return "<#= enumName #>";
+<#+
+    }
+#>
+                default:
+                    Debug.Fail("The enum value is not defined, please check the ExceptionArgument enum in ExceptionArgument.cs.");
+                    return "";
+            }
+        }
+    }
+<#+
+}
+
+//
+// Writes ThrowHelper.GetResourceString
+//
+
+void WriteThrowHelperGetResourceString(Type exceptionResourceType)
+{
+#>
+    internal static partial class ThrowHelper
+    {
+        private static string GetResourceString(ExceptionResource resource)
+        {
+            string resourceName;
+
+            // nameof(...) is used in the switch statement below to guarantee there's a matching property on the SR class.
+
+            switch (resource)
+            {
+<#+
+    foreach (string enumName in Enum.GetNames(exceptionResourceType))
+    {
+#>
+                case ExceptionResource.<#= enumName #>:
+                    resourceName = nameof(System.SR.<#= enumName #>);
+                    break;
+<#+
+    }
+#>
+                default:
+                    Debug.Fail("The enum value is not defined, please check the ExceptionResource enum in ExceptionResource.cs.");
+                    return "";
+            }
+
+            return System.SR.GetResourceString(resourceName, null);
+        }
+    }
+<#+
+}
+
+//
+// Writes the auto-gen warning
+//
+
+void WriteAutogenWarning(string filename)
+{
+#>
+!! THIS FILE IS AUTO-GENERATED !!
+// Any changes you make directly to this file will be overwritten by build tooling.
```

Comment:
What is the build tooling that will generate these files?

Reply:
I used VS because it was simplest. Make a change to the file, save, and VS auto-updates the code behind. Vector.tt and other files work the same way.

Reply:
Then the docs should say that you need to have the `System.Private.CoreLib.csproj` project opened and edit the files in VS. It is really VS that regenerates these files for you, not a build tooling.

What should folks without VS do? (e.g. on Linux)

Reply:
Folks without VS can continue to use the same workflow they do today: update the the enum + switch statement manually. (The definitions of these have been moved to the respective code-behind files rather than as part of ThrowHelper.cs.) They simply won't get the convenience of being able to update just the enum and have everything else auto-genned on their behalf.

The next time somebody does make a change to the T4 template, it'll re-gen the codebehinds. If something was missed or is out-of-date it'll immediately show up as a compilation error and can be addressed by whoever is making the change at that time.

Reply:
> Folks without VS can continue to use the same workflow they do today

Is there any MSBuild task that could just do the generation automatically as part of the building of the .csproj?

Reply:
I'm not aware of anything built-in. There are some third-party tools that can do this; see https://notquitepure.info/2018/12/12/T4-Templates-at-Build-Time-With-Dotnet-Core/ for an example.

Reply:
I guess without the tooling it's no worse than what we have today (other than all the extra gook surrounding it)?  I'm not sure it's worth the complexity, but I don't have a strong opinion.